### PR TITLE
Logstash code improvements

### PR DESCRIPTION
**Pull request by bperian**
* Minimized socket operations 
* Buffer cleanup on each dequeue 
* Better error handling

### CODE REVIEW

File: doc/README.logstash [hunk 1]
```diff
@@ -13,6 +13,73 @@ $ ntopng -F "logstash;localhost;tcp;5510"
 Syntax:
 logstash;<host>;<protocol>;<port_number>
 
+ntopng will NOT create the mappings automatically - a template should be created.
```
Comment (simonemainardi) [inline doc/README.logstash]: Why you keep putting this Elastic Search template configuration? This is not necessary in logstash. Please remove it.

Also note that, with ES export enabled, this index template is automatically pushed to ES.

File: src/Logstash.cpp [hunk 2]
```diff
@@ -174,78 +196,124 @@ void Logstash::sendLSdata() {
...
       elkExportedFlows += num_flows;
-      close(sockfd);
 
     } else {
       sleep(1);
     }
   } /* while */
+
+  close(sockfd);
+  if(proto)  free(proto);
```
Comment (simonemainardi) [inline src/Logstash.cpp]: why you put those two free back? I removed them as they can lead to segfault.

File: src/Logstash.cpp [hunk 3]
```diff
@@ -146,8 +146,30 @@ void Logstash::sendLSdata() {
...
   int portno;
+  int retval;
+  int skipDequeue = 0;
+  int sent = 0;
+  size_t sentLength = 0;
+
+  server = gethostbyname(ntop->getPrefs()->get_ls_host());
+  portstr = ntop->getPrefs()->get_ls_port();
+
+  if(server == NULL || portstr==NULL){
```
Comment (simonemainardi) [inline src/Logstash.cpp]: please, honour indentation: `portstr == NULL`

File: src/Logstash.cpp [hunk 4]
```diff
@@ -146,8 +146,30 @@ void Logstash::sendLSdata() {
...
+     //can't send
+     return;
+  }
+
+  proto = ntop->getPrefs()->get_ls_proto();
+  if(proto&&!strncmp(proto,"udp",3)){
```
Comment (simonemainardi) [inline src/Logstash.cpp]: indentation, missing space after comma and to separate &&

**Review by simonemainardi — CHANGES_REQUESTED** [review]
thanks for submitting this contribution. Please check and review according to comments.

**Comment by bperian** [comment]
Thanks Simone. Updated code in https://github.com/ntop/ntopng/pull/1143/commits/2b84ef0a68bfe478ef83207897a69810ff01c888

The old code resulted from the merge I did, sorry about that.

**Comment by simonemainardi** [comment]
thank you

**Pull request merged**